src/libmain/shared.cc

```diff
@@ -194,9 +195,26 @@ void initNix()
     /* HACK: on darwin, we need can’t use sigprocmask with SIGWINCH.
      * Instead, add a dummy sigaction handler, and signalHandlerThread
      * can handle the rest. */
-    struct sigaction sa;
-    sa.sa_handler = sigHandler;
-    if (sigaction(SIGWINCH, &sa, 0)) throw SysError("handling SIGWINCH");
+    act.sa_handler = sigHandler;
+    if (sigaction(SIGWINCH, &act, 0)) throw SysError("handling SIGWINCH");
+
+    /* Disable SA_RESTART for interrupts, so that system calls on this thread
+     * error with EINTR like they do on Linux.
+     * Most signals on BSD systems default to SA_RESTART on, but Nix
+     * expects EINTR from syscalls to properly exit. */
+    act.sa_handler = SIG_DFL;
+    if (sigaction(SIGINT, &act, 0)) throw SysError("handling SIGINT");
+    if (sigaction(SIGTERM, &act, 0)) throw SysError("handling SIGTERM");
+    if (sigaction(SIGHUP, &act, 0)) throw SysError("handling SIGHUP");
+    if (sigaction(SIGPIPE, &act, 0)) throw SysError("handling SIGPIPE");
+    if (sigaction(SIGQUIT, &act, 0)) throw SysError("handling SIGQUIT");
+    if (sigaction(SIGILL, &act, 0)) throw SysError("handling SIGILL");
+    if (sigaction(SIGTRAP, &act, 0)) throw SysError("handling SIGTRAP");
+    if (sigaction(SIGABRT, &act, 0)) throw SysError("handling SIGABRT");
+    if (sigaction(SIGFPE, &act, 0)) throw SysError("handling SIGFPE");
+    if (sigaction(SIGBUS, &act, 0)) throw SysError("handling SIGBUS");
+    if (sigaction(SIGXCPU, &act, 0)) throw SysError("handling SIGXCPU");
+    if (sigaction(SIGXFSZ, &act, 0)) throw SysError("handling SIGXFSZ");
```

Comment:
Some of these seem unnecessary, namely `SIGILL`, `SIGABRT`, `SIGFPE`, `SIGBUS`, `SIGXCPU` and `SIGXFSZ`. Really we only care about getting EINTR for a ctrl-c or client disconnect.